### PR TITLE
Fix docker-ce pinning integration-tests-runner and deprecated HDFS package.

### PR DESCRIPTION
**Pull request by qoega**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Docker in integration tests runner is updated to latest version. It was previously pinned u until patch release 24.0.3 was out. https://github.com/moby/moby/issues/45770#issuecomment-1618255130.
- HDFS image was deprecated and not running with current docker version. Switched to newer version of a derivative image based on ubuntu.
- HDFS tests were hardened to allow them to run with python-repeat.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 340214a246cd1c35d96cfb21be0576d87e05fea0 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66867/340214a246cd1c35d96cfb21be0576d87e05fea0/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Builds</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66868/e708219f6aecfe9934827f453665e03142ad5112/builds/report.html">✅ success</a></td></tr>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66806/e74892bfaf4aa98654a6253a34bbeb3f1740de41/docs_check.html">✅ success</a></td></tr>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/77c253634a1d45790e0a7429cfca476b1f93f07d/fast_test.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/77c253634a1d45790e0a7429cfca476b1f93f07d/unit_tests__release_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: docker/test/integration/runner/Dockerfile [hunk 1]
```diff
@@ -43,13 +43,11 @@ ENV TZ=Etc/UTC
...
 
 ENV DOCKER_CHANNEL stable
-# Unpin the docker version after the release 24.0.3 is released
-# https://github.com/moby/moby/issues/45770#issuecomment-1618255130
 RUN curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add - \
     && add-apt-repository "deb https://download.docker.com/linux/ubuntu $(lsb_release -c -s) ${DOCKER_CHANNEL}" \
     && apt-get update \
     && env DEBIAN_FRONTEND=noninteractive apt-get install --yes \
-        docker-ce='5:23.*' \
+        docker-ce \
```
Comment (alesapin) [inline docker/test/integration/runner/Dockerfile:50]: please pin it back to some recent version

Comment (qoega) [reply docker/test/integration/runner/Dockerfile:50]: Let's try this
```suggestion
        docker-ce="5:27.0.3*" \
```

**Review by alesapin — APPROVED** [review]

File: docker/test/integration/runner/Dockerfile (hunk 1 above)
Comment (antaljanosbenjamin) [reply docker/test/integration/runner/Dockerfile]: It is interesting, I had the same thing. However once I used `tests/integration/runner` and issued `docker version` from an integration test, I got the following:

```
Client: Docker Engine - Community
 Version:           24.0.5
 API version:       1.42 (downgraded from 1.43)
 Go version:        go1.20.6
 Git commit:        ced0996
 Built:             Fri Jul 21 20:35:18 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.6
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.9
  Git commit:       9dbdbd4
  Built:            Fri May  5 21:18:13 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.22
  GitCommit:        8165feabfdfe38c65b599c4993d227328c231fca
 runc:
  Version:          1.1.8
  GitCommit:        v1.1.8-0-g82f18fe
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

I have 0 idea what doesn server/client means in context of docker, but I think this pins the server version, not the client version.

**Comment by qoega** [comment]
So errors are actual incompatibilities of our tests with newer docker version! We need to fix them before merge...

**Comment by qoega** [comment]
Now I know why this happens, but I did not find the right path how we resolve host for client ip.

Here is an example. Client has 2 ips: `10.0.0.2` and `2001:3984:3989::1:1112` in test, but we get ipv4 mapped to ipv6 instead. 
Coredns does not resolve this one to ipv4 and I'm not sure if it should. I expect we should be able to resolve ipv4 address when we have mapped one.
```
coredns_1            | [INFO] 10.0.0.3:52760 - 38647 "PTR IN 2.0.0.0.0.0.a.0.f.f.f.f.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.ip6.arpa. udp 90 false 512" NOERROR qr,aa,rd 274 0.000116371s
```
We do have a helper in poco to detect those, but I'm still looking where exactly we resolve client address.
```
  bool isIPv4Mapped() const;
  /// Returns true iff the address is an IPv4 mapped IPv6 address.
  ///
  /// For IPv4 addresses, this is always true.
  ///
  /// For IPv6, the address must be in the ::FFFF:x:x range.
```

```
2024.07.29 16:57:21.109318 [ 13 ] {} <Debug> HTTP-Session: d13be957-7fd0-46e8-9287-ef8d6cf317cc Authenticating user 'test_dns' from [::ffff:10.0.0.2]:43816
2024.07.29 16:57:21.109958 [ 13 ] {} <Warning> AddressPatterns: Failed to check if the allowed client hosts contain address ::ffff:10.0.0.2. DB::Exception: ::ffff:10.0.0.2 c
ould not be resolved, code = 198
2024.07.29 16:57:21.136100 [ 13 ] {} <Error> Access(user directories): from: ::ffff:10.0.0.2, user: test_dns: Authentication failed: Code: 195. DB::Exception: Connections fr
om ::ffff:10.0.0.2 are not allowed. (IP_ADDRESS_NOT_ALLOWED), Stack trace (when copying this message, always include the lines below):

0. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/exception:141: Poco::Exception::Exception(String const&, int) @ 0x000000001612ec12
1. /home/ubuntu/clickhouse/src/Common/Exception.cpp:110: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c4d1c39
2. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/string:1499: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000006e2f16c
3. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/vector:438: DB::Exception::Exception<String>(int, FormatStringHelperImpl<std::type_identity<String>::type>, St
ring&&) @ 0x0000000006e2ed6b
4. /home/ubuntu/clickhouse/src/Access/IAccessStorage.cpp:0: DB::IAccessStorage::throwAddressNotAllowed(Poco::Net::IPAddress const&) @ 0x000000001060f8f4
5. /home/ubuntu/clickhouse/src/Access/IAccessStorage.cpp:531: DB::IAccessStorage::authenticateImpl(DB::Credentials const&, Poco::Net::IPAddress const&, DB::ExternalAuthentic
ators const&, bool, bool, bool) const @ 0x000000001060f82d
6. /home/ubuntu/clickhouse/src/Access/IAccessStorage.cpp:508: DB::IAccessStorage::authenticate(DB::Credentials const&, Poco::Net::IPAddress const&, DB::ExternalAuthenticator
s const&, bool, bool, bool) const @ 0x000000001060f5ff
7. /home/ubuntu/clickhouse/src/Access/MultipleAccessStorage.cpp:451: DB::MultipleAccessStorage::authenticateImpl(DB::Credentials const&, Poco::Net::IPAddress const&, DB::Ext
ernalAuthenticators const&, bool, bool, bool) const @ 0x000000001063d2d5
8. /home/ubuntu/clickhouse/src/Access/IAccessStorage.cpp:496: DB::IAccessStorage::authenticate(DB::Credentials const&, Poco::Net::IPAddress const&, DB::ExternalAuthenticator
s const&, bool, bool) const @ 0x000000001060f567
9. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:801: DB::AccessControl::authenticate(DB::Credentials const&, Poco::Net::IPAddress const&
, String const&) const @ 0x000000001058e7b9
10. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/string:1499: DB::Session::authenticate(DB::Credentials const&, Poco::Net::SocketAddress const&) @ 0x000000001
14ba879
11. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/string:1499: DB::authenticateUserByHTTP(DB::HTTPServerRequest const&, DB::HTMLForm const&, DB::HTTPServerResp
onse&, DB::Session&, std::unique_ptr<DB::Credentials, std::default_delete<DB::Credentials>>&, std::shared_ptr<DB::Context const>, std::shared_ptr<Poco::Logger>) @ 0x00000000
1299ff68
12. /home/ubuntu/clickhouse/src/Server/HTTPHandler.cpp:186: DB::HTTPHandler::authenticateUser(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&) @ 0x00000000129
28fe7
13. /home/ubuntu/clickhouse/src/Server/HTTPHandler.cpp:202: DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Ou
tput&, std::optional<DB::CurrentThread::QueryScope>&, StrongTypedef<unsigned long, ProfileEvents::EventTag> const&) @ 0x0000000012929362
14. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:290: DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&, St
rongTypedef<unsigned long, ProfileEvents::EventTag> const&) @ 0x000000001292e991
15. /home/ubuntu/clickhouse/base/poco/Foundation/include/Poco/AutoPtr.h:205: DB::HTTPServerConnection::run() @ 0x0000000012999c58
16. /home/ubuntu/clickhouse/base/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x00000000161ca9e7
17. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: Poco::Net::TCPServerDispatcher::run() @ 0x00000000161caebe
18. /home/ubuntu/clickhouse/base/poco/Foundation/src/ThreadPool.cpp:219: Poco::PooledThread::run() @ 0x000000001617d0b2
19. /home/ubuntu/clickhouse/base/poco/Foundation/include/Poco/AutoPtr.h:77: Poco::ThreadImpl::runnableEntry(void*) @ 0x000000001617ad63
20. ? @ 0x00007f5602ad9ac3
21. ? @ 0x00007f5602b6b850
 (version 24.8.1.1)
```

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**